### ISSUE TITLE
InVEST 3.8.9 won't launch on Big Sur

### ISSUE THREAD

**Issue by phargogh** [natcap/invest#383]
## To reproduce

1. Download InVEST 3.8.9
2. Unzip the archive
3. Do the right-click, open dance to get the application to launch
4. Observe in the dock that the icon bounces (indicating application load) and then closes (indicating application crash)

I then ran the application on the CLI with this output:

```shell
(base) ➜  InVEST-3.8.9 ./InVEST.app/Contents/MacOS/InVEST
Traceback (most recent call last):
  File "src/natcap/invest/cli.py", line 13, in <module>
ImportError: attempted relative import with no known parent package

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "src/natcap/invest/cli.py", line 18, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/usr/local/miniconda/envs/mac-env/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 627, in exec_module
  File "site-packages/natcap/invest/__init__.py", line 8, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/usr/local/miniconda/envs/mac-env/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 627, in exec_module
  File "site-packages/pygeoprocessing/__init__.py", line 13, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/usr/local/miniconda/envs/mac-env/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 627, in exec_module
  File "site-packages/pygeoprocessing/geoprocessing.py", line 53, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/usr/local/miniconda/envs/mac-env/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 627, in exec_module
  File "site-packages/shapely/wkb.py", line 4, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/usr/local/miniconda/envs/mac-env/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 627, in exec_module
  File "site-packages/shapely/geos.py", line 113, in <module>
  File "site-packages/shapely/geos.py", line 56, in load_dll
OSError: Could not find lib c or load any of its variants [].
```

**Comment by emlys** [comment]
Would this possibly be related to this note from the [changelog](https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-release-notes)?

> New in macOS Big Sur 11.0.1, the system ships with a built-in dynamic linker cache of all system-provided libraries. As part of this change, copies of dynamic libraries are no longer present on the filesystem. Code that attempts to check for dynamic library presence by looking for a file at a path or enumerating a directory will fail. Instead, check for library presence by attempting to dlopen() the path, which will correctly check for the library in the cache. (62986286)

**Comment by emlys** [comment]
If that is indeed the issue, it looks like it's fixed in [this Python PR]https://github.com/python/cpython/pull/22855#issuecomment-726323376), and they haven't decided yet if it'll be in 3.9 only, or backported to previous versions.

**Comment by phargogh** [comment]
After some digging, I'm convinced that this shared-library issue with the linker cache _is_ affecting the behavior of shapely, but it looks like the bug itself is in shapely ([quoted from geos.py](https://github.com/Toblerity/Shapely/compare/1.7.0...1.7.1#diff-449c80fbb0cac229a562aa3662da054b7ccdd3533e46d7e66cdf623fba6dd507R138-R142)) :

```diff
-    free = load_dll('c').free
+    # ctypes.CDLL(None) internally calls dlopen(NULL), and as the dlopen
+    # manpage says, "If filename is NULL, then the returned handle is for the
+    # main program". This way we can let the linker do the work to figure out
+    # which libc Python is actually using.
+    free = CDLL(None).free
```

To reproduce the issue on a Mac with Big Sur installed:

```python
$ conda create -p ./env37-conda-geos-shapely-1.7.0 -c conda-forge "shapely==1.7.0" "python=3.7"
(env37-conda-geos-shapely-1.7.0) $ python
Python 3.7.8 | packaged by conda-forge | (default, Jul 31 2020, 02:37:09)
[Clang 10.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
impor>>> import shapely.ops
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jdouglass/workspace/phargogh/invest/env37-conda-geos-shapely-1.7.0/lib/python3.7/site-packages/shapely/ops.py", line 13, in <module>
    from shapely.geos import lgeos
  File "/Users/jdouglass/workspace/phargogh/invest/env37-conda-geos-shapely-1.7.0/lib/python3.7/site-packages/shapely/geos.py", line 138, in <module>
    free = load_dll('c').free
  File "/Users/jdouglass/workspace/phargogh/invest/env37-conda-geos-shapely-1.7.0/lib/python3.7/site-packages/shapely/geos.py", line 62, in load_dll
    libname, fallbacks or []))
OSError: Could not find lib c or load any of its variants [].
```

Alteratively, here's shapely 1.7.1, which works fine:
```python
$ conda create -y -p ./env37-conda-geos-shapely-1.7.1 -c conda-forge "shapely==1.7.1" "python=3.7"
(env37-conda-geos-shapely-1.7.1) $ python
Python 3.7.8 | packaged by conda-forge | (default, Jul 31 2020, 02:37:09)
[Clang 10.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import shapely.geos
>>> import shapely.ops
```

As mentioned in https://github.com/natcap/invest/issues/200, this immediate issue is solved in the `release/3.9` branch by the upgrade to `shapely>=1.7.0`.  Since this issue was _actually_ fixed in Shapely 1.7.1, I'll test a patch for this and PR it in to `release/3.9`.

**Comment by phargogh** [comment]
I'll mark this issue as resolved, though the fix has been merged into `release/3.9` and we probably won't be releasing a bugfix release for 3.8.x.  Issue #384  is still outstanding, but this immediate crash has been resolved.
